---
title: "Report: Transcriptomics and Single Cell Genomics"
author: "Pedro Bueso-Inchausti Garcia"
date: "2020-1-13"
output:
   pdf_document:
      fig_caption: true
      number_sections: true
---

```{r, message=FALSE, echo=FALSE}

#We set the working directory.

setwd("C:/Users/User/Desktop/GDAV/Transcriptomics")

#We load the packages that will be needed.

options(warn=-1)
library(tidyr)
library(dplyr)
library(ggplot2)
library(gridExtra)
library(Seurat)
library(Matrix)
library(knitr)
library(kableExtra)

#We create the functions that will be needed.

feature_function = function(vector, search, feature1, feature2)
{
  vector[!(vector %in% search)]=feature2
  vector[vector %in% search]=feature1; return(vector)
}

label_function = function(vector, search)
{vector[!(vector %in% search)]=" "; return(vector)}
```

\newpage

\tableofcontents

\newpage

# Background

## PCA (linear dimensionality reduction)

When information is collected from a data sample, the most frequent is to take as many variables as possible. However, we need to be aware that some of them might be related or measure the same aspect from different points of views. If that is the case, it might be necessary to reduce the number of variables. 

The Principal Component Analysis (PCA) \cite{1} consists on the transformation of the set of original variables into another set, the Principal Components (PCs), obtained as a linear combination of the originals. The new variables, in the same number as the originals, retain all the variability. However, most of the PCs explain such a small variability that can be ignored, while a few PCs can be considered without a significant loss of information. PCs are independent between each other, so they explain the maximum possible residual variability that has not been explained by previous ones.

Before performing the PCA, there are a couple of things that need to be considered. The first thing is whether the original variables are correlated; only if they are does PCA makes sense. The second thing is whether the original variables are heterogeneous -expressed in different units of measure- or homogeneous -expressed in the same units of measure-. In the first case, it would be necessary to use the correlation matrix, which normalises all the variables; in the second case, using the covariance matrix, which implies no loss of information, would be more appropiate. 

Once the PCA has been performed, there are some metrics that should be examined. The eigenvalues are the variance explained by each PC. The eigenvectors are the linear combinations that define each PC. The loadings are the eigenvector coefficients; this is, the degree in which each original variable influences a PC. The scores are the original instances put into the PCs units. All these metrics can be visualized for better understanding. The sediment graph shows the eigenvalues for each PC; although there are other methods, it is quite common for determining the how many to use (note that a high number of PCs can explain a greater proportion of the total variability while a low number allows for greater simplicity in the representation). The loading plot shows, for the first two PCs, the loadings of each variable; the variables furthest from 0 in the horizontal axis are the ones giving meaning to the first PC; the variables furthest from 0 in the vertical axis are the ones giving meaning to the second PC; the variables close to the centre or displaced in both axis do not help in the interpretation of the PCs. The score plot shows, for the first two PCs, the scores of each instance; this representation allows describing, in a simple way, the multidimensional dataset.

Although PCA is an interesting technique when dealing with complex datasets, it has some drawbacks. The first is the partial loss of information. The second is that, while the original variables have real meaning, the PCs generally lack such meaning; therefore, the interpretability of the results is worst.

## t-SNE and UMAP (nonlinear dimensionality reduction)

Nonlinear dimensionality reduction techniques allow embedding high-dimensional data in a low-dimensional space, which is appropriate for visualization. This is possible because they convert similar high-dimensional instances into nearby points and dissimilar high-dimensional instances into distant points. How is this done? The mathematics are a bit advance but the intuition behind is simple. They first built a high-dimensional representation of the data that establishes which instances are connected; they then optimize a low-dimensional representation to be as structurally similar as possible.

T-distributed Stochastic Neighbor Embedding (t-SNE) \cite{2} was the original technique. Its performance, however, suffers with large datasets. Uniform Manifold Approximation and Projection (UMAP) \cite{3} is a new technique that is faster, scaling well in terms of size and dimensionality, and allows preserving not only the local structure of the data (whether two instances belong to the same cluster) but also its global structure (whether two clusters are more similar between them than other two). One big problem for both techniques is that using them correctly can be challenging as clusters are influenced by the chosen parametrization.

\newpage

# Practice 1: PCA for studying genetic expression on cell lines

## Objective

In this practice, we will study the genetic expression of different cell lines. We have information for a wild-type, a mutant and complemented lines (mutants to which we add a transcription factor). By adding these TFs, we expect to complement the loss of genetic expression in the mutant; some TFs will achieve a total complementation while others will undercomplement or overcomplement the wild-type expression. Some complemented lines might even develop new functionalities. PCA will be used to explore such hypothesis.

## Data preparation

The first thing we do is to read the table with the gene expression and keep the subsets of interest: first_subset (includes genes, mutant, wild-type and complemented lines BLJ, JKD, MGP, NUC, IME, SCR), second_subset (adds the artificially created transcriptomes with 25% wt + 75% mut, 50% wt + 50% mut, 75% wt + 25% mut) and third_subset (adds the SCR domain).

```{r, echo=FALSE}
data = read.csv("table.csv", row.names=1)
first_subset = data[,1:8]
second_subset = transform(first_subset,
                        comp25=0.25*first_subset$J0571+0.75*first_subset$shrJ0571,
                        comp50=0.50*first_subset$J0571+0.50*first_subset$shrJ0571,
                        comp75=0.75*first_subset$J0571+0.25*first_subset$shrJ0571)
third_subset = transform(first_subset, SCRdomain=data$SCRdomain)
first_matrix = data.matrix(first_subset)
second_matrix = data.matrix(second_subset)
third_matrix = data.matrix(third_subset)
```

## Analyze transcriptome differences between cell lines

The relationships between cell lines are calculated in terms of disimilarity. As it is very difficult to perform comparisons based on thousands of genes, we need to do a dimensionality reduction. This is achieved through PCA, which transforms thousands of correlated genes into a few uncorrelated PCs.

```{r, echo=FALSE}
first_PCA = princomp(first_matrix, cor=FALSE, scores=TRUE)
variance = first_PCA$sdev/sum(first_PCA$sdev)
scores = first_PCA$scores
loadings = first_PCA$loadings
```

By plotting the eigenvalues associated to each PC, we can see how they contribute to the variance of the original set. In this case, there are few PCs that contribute to most of the variance. From now on, we will be working with the first 2, that accumulate almost 90% of it.

```{r, out.width="70%", fig.align="center", echo=FALSE}
df = data.frame(variance); df=tibble::rownames_to_column(df, "PCs")
ggplot(df, aes(x=PCs,y=variance)) + geom_col()
```

\newpage

By plotting the loadings, we can see which cell lines are closer to others (we focuse primaly on the first PC, this is, the X axis). The complemented lines JKD, SCR and NUC are closer to the wild-type, which means that they were closer than others to a total complementation. The complemented lines MGP, BLJ and IME are closer to the mutant, which means that they were further from a total complementation.

```{r, out.width="70%", fig.align="center", echo=FALSE}
df = data.frame(PC1=loadings[,1], PC2=loadings[,2])
df = tibble::rownames_to_column(df, "cell_lines")
df$position = c(-0.01,0.01,-0.015,0.015,-0.015,0.015,-0.015,0.015)
ggplot(df, aes(x=PC1,y=PC2,label=cell_lines))+ geom_point() + geom_text(nudge_x=df$position)
```

## Repeat but consider artificial genomes

As the level of complementation increases, the artificial genomes appear closer to the wild-type and further from the mutant. It is interesting to see how such changes happen in a completely linear way.

```{r, echo=FALSE}
second_PCA = princomp(second_matrix,cor=FALSE,scores=TRUE)
loadings = second_PCA$loadings
```

```{r, out.width="70%", fig.align="center", echo=FALSE}
genomes = c("shrJ0571", "J0571", "comp25", "comp50", "comp75")
df = data.frame(PC1=loadings[,1], PC2=loadings[,2])
df = tibble::rownames_to_column(df, "cell_lines")
df$color = feature_function(df$cell_lines, genomes, "red", "black")
ggplot(df, aes(x=PC1,y=PC2,label=cell_lines)) + 
geom_point(color=df$color) + geom_text(nudge_y=-0.03,color=df$color)
```

\newpage

## Repeat but consider SCR domain

The complemented lines JKD (JACKDAW) and SCR (SCARECROW) appear closer to the SCR domain, which does not come as a surprise to us. It has been studied that JKD and SCR form complexes which are involved in the regulation of the stem cell fate; this is, how a particular cell develops into a final cell type \cite{4}. Therefore, one would expect that they appear as related.

```{r, echo=FALSE}
third_PCA = princomp(third_matrix,cor=FALSE,scores=TRUE)
loadings = third_PCA$loadings
scores = third_PCA$scores
```

```{r, out.width="60%", fig.align="center", echo=FALSE}
genomes = c("shrJ0571.JKD", "shrJ0571.SCR", "SCRdomain")
df = data.frame(PC1=loadings[,1],PC2=loadings[,2])
df = tibble::rownames_to_column(df, "cell_lines")
df$color = feature_function(df$cell_lines, genomes, "red", "black")
ggplot(df, aes(x=PC1,y=PC2,label=cell_lines)) + 
geom_point(color=df$color) + geom_text(nudge_x=0.02,color=df$color)
```

## Find the genes contributing to these transcriptomic changes

The normalised contribution of every gene is given by the PCA scores. We focuse only on the first PC, as it is the one which explain most of the variance. By sorting that column, we can retrieve the the genes with higher scores (both positive and negative). We create a matrix with the expression of these genes in each of the cell lines and represent that matrix with a heatmap. In the heatmap including the positive genes (the only one shown), we can see tendencies which are coherent with our previous analysis (the cell lines are ordered based on their degree of similarity).

```{r, echo=FALSE}
neg_genes = labels(head(sort(scores[,"Comp.1"]),20))
pos_genes = labels(tail(sort(scores[,"Comp.1"]),20))
neg_gene_exp=c(); for (gene in neg_genes){neg_gene_exp=rbind(neg_gene_exp,third_subset[gene,])}
pos_gene_exp=c(); for (gene in pos_genes){pos_gene_exp=rbind(pos_gene_exp,third_subset[gene,])}
```

```{r, out.width="70%", fig.align="center", echo=FALSE}
df = data.frame(pos_gene_exp)
df = tibble::rownames_to_column(df, "Genes")
df = gather(df, names(df[-1]), key="Cell_lines", value="Expression")
ggplot(df, aes(x=Cell_lines,y=Genes, fill=Expression)) + geom_tile() +  
scale_fill_gradient(low="darkblue",high="yellow") + theme_minimal() + 
theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1)) +
scale_x_discrete(limits=c("shrJ0571","shrJ0571.BLJ","shrJ0571.MGP","shrJ0571.IME","J0571",
                          "shrJ0571.NUC","shrJ0571.JKD","shrJ0571.SCR","SCRdomain"))
```

\newpage

# Practice 2: PCA for detecting biomarkers on different cell lines

## Objective

In this practice, we will study the genetic expression of different cell lines. In particular, we want to detect cell lines with a variable expression as well as to segregate ones from others based on biomarkers detection. To do so, we will be using PCA.

## Data preparation

The first thing we do is to read the table with the gene expression and keep the subsets of interest: first_subset (includes all the genes) and second_subset (includes the biomarkers for stem cells; these are genes which are expressed in WOX5 domain and not expressed in rest of cell lines).

```{r, echo=FALSE}
data=read.csv("TableS5.csv", row.names=1)
first_subset = data
second_subset = subset(first_subset,first_subset$WOX5>1 & 
                       rowSums(first_subset[,-which(names(first_subset)=="WOX5")]>1)<1)
first_matrix = data.matrix(first_subset)
second_matrix = data.matrix(second_subset)
second_names = rownames(second_subset)
```

## Visualize distribution of gene expression across cell lines

We use the boxplot as a way to show distributions. We can see that both E30 and S18 show a greater variance in their expression. They are very specialised cell lines, which means that some genes express little while others express a lot. This could explain their high variance.

```{r, out.width="70%", fig.align="center", echo=FALSE}
df=data.frame(first_subset)
df=tibble::rownames_to_column(df, "Genes")
df=gather(df, names(df[-1]), key="Cell_lines", value="Expression")
ggplot(df, aes(x=Cell_lines,y=Expression)) + geom_boxplot()
```

\newpage

## Analyze transcriptome differences between cell lines

The relationships between cell lines are calculated in terms of disimilarity. As it is very difficult to perform comparisons based on thousands of genes, we need to perform a dimensionality reduction. This is achieved through PCA, which transforms thousands of correlated genes into a few uncorrelated PCs.

```{r, echo=FALSE}
first_PCA = princomp(first_matrix,cor=FALSE,scores=TRUE)
variance = first_PCA$sdev[1:9]/sum(first_PCA$sdev[1:9])
scores = first_PCA$scores
loadings = first_PCA$loadings
```

By plotting the eigenvalues associated to each PC, we can see how they contribute to the variance of the original set. In this case, there are quite a lot of PCs that contribute to the variance. From now on, we will be working with the first 4, that accumulate almost 70% of it.

```{r, out.width="70%", fig.align="center", echo=FALSE}
df = data.frame(variance)
df = tibble::rownames_to_column(df, "PCs")
ggplot(df, aes(x=PCs,y=variance)) + geom_col()
```

By plotting the loadings, we can see which cell lines are closer to others. The cell lines E30 and S18 are clearly segregated from the others, which means that their gene expression is very different. Note that we should give more importance to the first plot, as it results from the combination of the first two PCs.

```{r, out.width="70%", fig.align="center", echo=FALSE}
df = data.frame(PC1=loadings[,1], PC2=loadings[,2], PC3=loadings[,3], PC4=loadings[,4])
df = tibble::rownames_to_column(df, "cell_lines")
df$cell_lines = label_function(df$cell_lines, c("E30", "S18"))
df$color = feature_function(df$cell_lines, c("E30", "S18"), "red", "black")
p1 = ggplot(df, aes(x=PC1,y=PC2,label=cell_lines)) + 
geom_point(color=df$color) + geom_text(nudge_y=0.1,color=df$color)
p2 = ggplot(df, aes(x=PC3,y=PC4,label=cell_lines)) + 
geom_point(color=df$color) + geom_text(nudge_y=0.1,color=df$color)
grid.arrange(p1, p2, ncol=1)
```

\newpage

## Find biomarkers for stem cells 

By plotting the heatmap, we can verify that the expression of stem cell biomarkers is higher in WOX5 than in any other cell line. The biomarkers were filtered as genes which are expressed in WOX5 and not expressed in other cell lines; therefore, this is the result one would expect.

```{r, out.width="70%", fig.align="center", echo=FALSE}
df = data.frame(second_subset)
df = tibble::rownames_to_column(df, "Genes")
df = gather(df, names(df[-1]), key="Cell_lines", value="Expression")
ggplot(df, aes(x=Cell_lines,y=Genes, fill=Expression)) + geom_tile() +
scale_fill_gradient(low="darkblue",high="yellow") + theme_minimal() +
theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1)) + theme(axis.text.y=element_blank())
```

## Separate cell lines based on biomarkers

If we just consider the selected biomarkers, plotting the loadings shows how WOX5 is clearly separated from the other cell lines. Note that we should give more importance to the first plot, as it results from the combination of the first two PCs.

```{r, echo=FALSE}
second_PCA = princomp(second_matrix,cor=FALSE,scores=TRUE)
loadings = second_PCA$loadings
```

```{r, out.width="70%", fig.align="center", echo=FALSE}
df = data.frame(PC1=loadings[,1], PC2=loadings[,2], PC3=loadings[,3], PC4=loadings[,4])
df = tibble::rownames_to_column(df, "cell_lines")
df$cell_lines = label_function(df$cell_lines, c("WOX5"))
df$color = feature_function(df$cell_lines, c("WOX5"), "red", "black")
p1=ggplot(df, aes(x=PC1,y=PC2,label=cell_lines)) + 
geom_point(color=df$color) + geom_text(nudge_y=0.1,color=df$color)
p2=ggplot(df, aes(x=PC3,y=PC4,label=cell_lines)) + 
geom_point(color=df$color) + geom_text(nudge_y=0.1,color=df$color)
grid.arrange(p1, p2, ncol=1)
```
\newpage

## Visualize biomarkers in the PCA scores

Now that we have seen how the selected biomarkers segregate WOX5 from the rest of the cell lines, it is interesting to see whether these genes can be identified as a separated group in a score plot. What we see, by plotting the scores, is that the selected biomarkers (in red) group together, However, and although they have extreme values for the first PC, they are not clearly segregated from the rest of genes.

```{r, out.width="70%", fig.align="center", echo=FALSE}
df = data.frame(PC1=scores[,1], PC2=scores[,2], PC3=scores[,3], PC4=scores[,4])
df$color = feature_function(rownames(df), second_names, "red", "black")
df$size = as.numeric(feature_function(rownames(df), second_names, 5, 1))
p1 = ggplot(df, aes(x=PC1,y=PC2)) + geom_point(color=df$color, size=df$size)
p2 = ggplot(df, aes(x=PC3,y=PC4)) + geom_point(color=df$color, size=df$size)
grid.arrange(p1, p2, ncol=1)
```

\newpage

# Practice 3: t-SNE for building transcriptome atlas map

## Objective

In this practice, we will study the genetic expression of single-cells coming from a planaria sample. The main goal is to build a cell line transcriptome atlas map. To do so, we will use t-SNE and UMAP.

## Data preparation

The first thing we do is to read the data and convert it into a sparse matrix (which result in significant memory and speed savings when many cells are zeros). From such matrix, we create a seurat object, container for both data and analysis (we create two objects, according to what is specified in excercises 1 and 2). Their differences refer to the number of cells in which a gene has to be expressed for it to be considered (1 in the first object vs 3 in the second) and to the number of genes which have to be expressed in a cell for it to be considered (1 in the first object vs 200 in the second).

```{r, echo=FALSE}
data = read.table("SCOplanaria.txt", row.names=1)
matrix = as.matrix(data)
sparse_matrix = Matrix(data=matrix, sparse=TRUE)
seurat1 = CreateSeuratObject(counts=sparse_matrix, project="SCOP1", min.cells=1, min.features=1)
seurat2 = CreateSeuratObject(counts=sparse_matrix, project="SCOP2", min.cells=3, min.features=200)
```

## Quality control, normalization, features selection and scaling

We explore the data with metrics. As expected, the number of genes and cells in the second object is lower.

```{r, echo=FALSE}
seurat1_num_genes = dim(seurat1)[1]
seurat1_num_cells = dim(seurat1)[2]
seurat1_genes_per_cell = mean(seurat1@meta.data$nFeature_RNA)
seurat1_reads_per_cell = mean(seurat1@meta.data$nCount_RNA)
seurat2_num_genes = dim(seurat2)[1]
seurat2_num_cells = dim(seurat2)[2]
seurat2_genes_per_cell = mean(seurat2@meta.data$nFeature_RNA)
seurat2_reads_per_cell = mean(seurat2@meta.data$nCount_RNA)

metrics = data.frame(
  seurat1 = c(seurat1_num_genes,seurat1_num_cells,seurat1_genes_per_cell,seurat1_reads_per_cell),
  seurat2 = c(seurat2_num_genes,seurat2_num_cells,seurat2_genes_per_cell,seurat2_reads_per_cell), 
  row.names = c("Total number of genes", "Total number of cells", 
  "Average number of genes per cell", "Average number of reads per cell"))
kable(metrics,digits=3) %>% kable_styling(font_size=10)
```

There is a clear correlation between number of genes (nFeature_RNA) and reads (nCount_RNA) per cell.

```{r, out.width="70%", fig.align="center", echo=FALSE}
p1 = FeatureScatter(seurat1, feature1="nFeature_RNA", feature2="nCount_RNA") + NoLegend()
p2 = FeatureScatter(seurat2, feature1="nFeature_RNA", feature2="nCount_RNA") + NoLegend()
CombinePlots(plots = list(p1, p2))
```

We perform a quality control. Low quality cells or empty droplets often have very few genes; cell doublets or multiplets often have to many genes. Thus, we keep those cells with not to few and not to many genes. It looks reasonable to filter cells with a number of genes between 200 and 2,500.

```{r, echo=FALSE}
seurat1 = subset(seurat1, subset=nFeature_RNA>200&nFeature_RNA<2500)
seurat2 = subset(seurat2, subset=nFeature_RNA>200&nFeature_RNA<2500)
```

\newpage

We perform the normalization, using the LogNormalization method. This normalizes the gene expression for each cell by the total expression, multiplies by a scale factor and log-transforms the result. If we compare the violin plots pre-normalization (not shown) and post-normalization, the normalization of the data is made evident (not all the point are grouped on the lower part of the plot).

```{r, out.width="70%", fig.align="center", echo=FALSE}
seurat1 = NormalizeData(seurat1, normalization.method="LogNormalize", scale.factor=10000)
seurat1 = AddMetaData(seurat1, colSums(seurat1[["RNA"]]@data), col.name="Normalised_nCount_RNA")
seurat2 = NormalizeData(seurat2, normalization.method="LogNormalize", scale.factor=10000)
seurat2 = AddMetaData(seurat2, colSums(seurat2[["RNA"]]@data), col.name="Normalised_nCount_RNA")

p1 = VlnPlot(seurat1, features="Normalised_nCount_RNA") + NoLegend()
p2 = VlnPlot(seurat2, features="Normalised_nCount_RNA") + NoLegend()
CombinePlots(plots = list(p1, p2))
```

We perform the feature selection by identifying the subset of genes whose expression varies the most from cell to cell. Focusing on these genes helps to highlight biological signals in single-cell datasets. We consider 2000 genes for the first seurat object and 3000 genes for the second seurat object (in red). The 2 most variable appear labelled.

```{r, message=FALSE, out.width="70%", fig.align="center", echo=FALSE}
seurat1 = FindVariableFeatures(seurat1, selection.method="vst", nfeatures=2000)
seurat2 = FindVariableFeatures(seurat2, selection.method="vst", nfeatures=3000)

p1 = LabelPoints(plot=VariableFeaturePlot(seurat1), 
points=head(VariableFeatures(seurat1),2), repel=TRUE) + NoLegend()
p2 = LabelPoints(plot=VariableFeaturePlot(seurat2), 
points=head(VariableFeatures(seurat2),2), repel=TRUE) + NoLegend()
CombinePlots(plots = list(p1, p2))
```

We perform the scaling by applying a linear transformation that makes mean expression and variance across cells 0 and 1. This is a way of regressing out variability (without specifying the source of variation removed).

```{r, message=FALSE, echo=FALSE}
seurat1 = ScaleData(seurat1, features=rownames(seurat1))
seurat2 = ScaleData(seurat2, features=rownames(seurat1))
```

\newpage

## PCA, t-SNE and UMAP

We perform a linear dimension reduction with PCA.

```{r, message=FALSE, echo=FALSE}
seurat1 = RunPCA(seurat1, features=VariableFeatures(object=seurat1))
seurat2 = RunPCA(seurat2, features=VariableFeatures(object=seurat2))
```

Seurat clusters the cells based on PCA scores, with each PC representing a metagene that combines information across correlated gene sets. To identify the dimensionality of the dataset, there are two alternative methods. The JackStrawPlot function provides a comparison of the distribution of p-values for each PC with a uniform distribution; significant PCs show a strong enrichment of genes with low p-values. The ElbowPlot provides a ranking of PCs based on the percentage of variance explained by each one. Although we don't show this plots, they were examined and 10 PCs look like a good representation of the original dataset.

We visualize the PCA results with a heatmap, where cells and genes are ordered according to their scores. This allows for easy exploration of the primary sources of heterogeneity in a dataset. It can be useful when trying to decide which PCs to include in further downstream analyses. Although we just show the results for the first PC, 10 have been examined.

```{r, out.width="70%", fig.align="center", echo=FALSE}
DimHeatmap(seurat1, dims=1, cells=500, balanced = TRUE)
DimHeatmap(seurat2, dims=1, cells=500, balanced = TRUE)
```

\newpage

We perform a clustering analysis based on the previously identified significant PCs. This implies constructing a K-nearest neighbor graph and then clustering the cells through modularity optimization techniques. For the first object, we use 10 PCs and a resolution of 0.5; for the second object, we use 5 PCs and a resolution of 0.6.

```{r, message=FALSE, results="hide", echo=FALSE}
seurat1 = FindNeighbors(seurat1, dims=1:10)
seurat1 = FindClusters(seurat1, resolution=0.5)
seurat2 = FindNeighbors(seurat2, dims=1:5)
seurat2 = FindClusters(seurat2, resolution=0.6)
```

We use UMAP and t-SNE to visualize the data in low-dimensional graphs. Only UMAP results are shown. We chose this technique because it appears to give better results. In the preparation of the seurat objects, we set different initial conditions (beings more restrictive in the second object) and we did a different feature selection (being more restrictive in the first objects).In the cluster search, the only differences between both seurat objects was the number of PCs and the resolution. Increasing the resolution can lead to having larger clusters (the second object has less clusters of a bigger size). Decreasing the number of PCs used can lead to overlapping between the clusters (the second object has more overlapping between clusters than the first one). We will continue our analysis considering only the first seurat object.  

```{r, message=FALSE, out.width="80%", fig.align="center", echo=FALSE}
seurat1 = RunUMAP(seurat1, dims=1:10)
seurat1 = RunTSNE(seurat1, dims=1:10)
seurat2 = RunUMAP(seurat2, dims=1:5)
seurat2 = RunTSNE(seurat2, dims=1:5)

p1 = DimPlot(seurat1, reduction="umap", label=TRUE, pt.size=3) + NoLegend()
p2 = DimPlot(seurat2, reduction="umap", label=TRUE, pt.size=3) + NoLegend()
CombinePlots(plots = list(p1, p2))
```

\newpage

## Extract top biomarkers for each cluster

Biomarkers are genes that are expressed significantlly more (positive) or less (negative) in one cell line. We search for biomarkers comparing every cluster to the rest. The argument "min.pct" tests only those genes that are detected at a minimum percentage in the all cells; the argument "logfc.threshold" tests only those genes showing a minimum difference between cells lines. This table shows the first five biomarkers for each of the clusters previously identified.

```{r, message=FALSE, results="hide", echo=FALSE}
biomarkers = FindAllMarkers(seurat1, only.pos=TRUE, min.pct=0.25, logfc.threshold=0.25)
biomarkers_top5 = biomarkers %>% group_by(cluster) %>% top_n(n=5, wt=avg_logFC)
biomarkers_top10 = biomarkers %>% group_by(cluster) %>% top_n(n=10, wt=avg_logFC)
biomarkers_top5_table = biomarkers_top5 %>% group_by(cluster) %>% 
                        summarise(gene=paste(gene,collapse=', '))
kable(biomarkers_top5_table) %>% kable_styling(font_size=10)
```

We generate an expression heatmap for this biomarkers. This allows us to see how cells belonging to the same cluster share a common expression pattern. It is interesting to see that some clusters (1-5, 2-9 and 4-6) share expression patterns. They are probably corresponding to related cell lines. 

```{r, out.width="70%", fig.align="center", echo=FALSE}
DoHeatmap(seurat1, features=biomarkers_top5$gene) + NoLegend()
```

## Rename clusters based on expression distribution of markers

This is the given list of markers and the cell identities they are associated to.

```{r, echo=FALSE}
markers_identities = list(
  "dd-Smed-v6-61-0"="Early epidermal progenitors","dd-Smed-v6-2178-0"="Late epidermal progenitors",
  "dd-Smed-v6-298-0"="Epidermis","dd-Smed-v6-1410-0"="Muscle progenitors",
  "dd-Smed-v6-702-0"="Muscle body","dd-Smed-v6-2548-0"="Neural progenitors",
  "dd-Smed-v6-9977-0"="GABA neurons","dd-Smed-v6-48-0"="Phagocytes",
  "dd-Smed-v6-175-0"="Parenchymal cells","dd-Smed-v6-1161-1"="Pigment")
markers = names(markers_identities)
cell_lines = unname(unlist(markers_identities))
markers_identities_df = data.frame(markers,cell_lines)
kable(markers_identities_df) %>% kable_styling(font_size=10)
```

We can create the correspondance between cluster and cell identities with two methodologies: Method 1 (M1) uses VlnPlot and FeaturePlot, that trace the gene expression across clusters visually. Method 2 (M2) searches for the genes among the biomarkers already retrieved in previous steps.

```{r, echo=FALSE}
clusters = sort(unique(Idents(seurat1)))
clusters_identities_M1 = data.frame(clusters,
identities=c("Neural progenitors","Early epidermal progenitors","Not defined","Epidermis",
"Muscle body","Early epidermal progenitors","Muscle progenitors","Phagocytes","Not defined",
"GABA neurons","Pigment","Late epidermal progenitors","Parenchymal cells"))

clusters_identities_M2 = data.frame(clusters, identities = rep("Not defined", length(clusters)))
clusters_identities_M2$clusters = as.character(clusters_identities_M2$clusters)
clusters_identities_M2$identities = as.character(clusters_identities_M2$identities)

for (marker in markers)
{
  cluster = filter(biomarkers_top10, gene==marker)$cluster
  cell_identity = markers_identities[[marker]]
  for (element in cluster)
  {clusters_identities_M2$identities[clusters_identities_M2$clusters==element]=cell_identity}
}
```

\newpage

Both methodologies give as the same information. Now that we have the correspondencies, we can see how clusters 1 and 5 refer to the same cell line (Early epidermal progenitors) and clusters 4 and 6 refer to very similar cell lines (Muscle body and Muscle progenitors). The thing with clusters 2 and 9 is that cluster 2 has not been defined; we can assume it will be a similar cell line to GABA neurons, which is what cluster 9 stands for. 

```{r, echo=FALSE}
clusters_identities_M12 = data.frame(clusters, 
identities_M1=clusters_identities_M1$identities, identities_M2=clusters_identities_M2$identities)
kable(clusters_identities_M12) %>% kable_styling(font_size=10)
```

After including the cell identities names, we use UMAP and t-SNE to visualize the data in low-dimensional graphs. Only UMAP results are shown in the report.

```{r, out.width="70%", fig.align="center", echo=FALSE}
new_clusters = clusters_identities_M2$identities
names(new_clusters) = levels(seurat1)
seurat1 = RenameIdents(seurat1, new_clusters)
DimPlot(seurat1, reduction="umap", label=TRUE, pt.size=3) + NoLegend()
```

If we compare it with the tree reconstruction of planarian cell lines shown in \cite{5}, it seems like most of the general cell lines appear. This makes sense because our dataset is a sample of planaria and, even if we had took a local biopsy, it is expected to find a mixture of cell lines which are not tissue specific. Our central cluster includes neural progenitors; in the linage tree reconstruction, however, it is neoblast that appears. The next exercise will prove that such cell line also exists in our dataset, having an important role in the central cluster. 

\newpage

## Show the distribution of a neoblast marker gene

Neoblasts are distributed all over the body and represent between 25-30% of all the cells. As they are not tissue specific, we can find one of its marker genes spread across the transcriptome atlas map (though it is more abundant in the central cluster).


```{r, out.width="70%", fig.align="center", echo=FALSE}
VlnPlot(seurat1, features="dd-Smed-v6-1999-0") + NoLegend()
FeaturePlot(seurat1, features="dd-Smed-v6-1999-0") + NoLegend()
```

\newpage

\begin{thebibliography}{9}
\bibitem{1}
"RPubs - Analisis de Componentes Principales (Principal Component Analysis, PCA) y t-SNE", Online, Available: \url{https://rpubs.com/Joaquin_AR/287787}, Accessed: 28-Dec-2019.
\bibitem{2}
"t-distributed stochastic neighbor embedding - Wikipedia", Online, Available: \url{https://en.wikipedia.org/wiki/T-distributed_stochastic_neighbor_embedding}, Accessed: 28-Dec-2019.
\bibitem{3}
"Understanding UMAP", Online, Available: \url{https://pair-code.github.io/understanding-umap/}, Accessed: 28-Dec-2019.
\bibitem{4}
"Locus: AT5G03150", Online, Available: \url{https://www.arabidopsis.org/servlets/TairObject?type=locus&name=AT5G03150}, Accessed: 28-Dec-2019.
\bibitem{5}
M. Plass et al., "Cell type atlas and lineage tree of a whole complex animal by single-cell transcriptomics", Science, vol. 360, no. 6391, May 2018.
\end{thebibliography}
